p4-16/spec/P4-16-spec.mdk

```diff
@@ -1460,6 +1460,123 @@ lvalue
 The following is a legal l-value: `headers.stack[4].field`.  Note
 that method and function calls cannot return l-values.
 
+## Named parameters {#sec-named-params}
+
+P4 supports named parameters to enable invoking constructors, controls,
+parsers, and externs with a subset of the arguments when appropriate default
+values are provided. Specifically:
+
+- Calls specify parameters by name. By default, all parameters must be
+  specified.
+- Optional parameters are specified by assigning a default parameter value:\
+  `param_modifier param_type name = default_parameter_value`
+- Mandatory parameters must preceed all optional parameters and all parameters
+  must be called in the order in which they were declared. This is similar to
+  C++, except that preceeding optional parameters can be fully omitted (as in
+  Python).
+- We only allow optional parameters for data-plane program entities that are
+  not exposed at runtime, i.e., do not participate in the generation of
+  control-plane APIs. These include:
+   - Constructors, including extern constructors and package constructors,
+     controls, and parsers
+   - Extern methods
+   - Internal actions or functions that are used exclusively in the data-plane
+- Optional parameters must be either `in` or directionless
+
+### Default values
+
+If `_` is provided as the default value the compiler will initialize the
+optional parameter with an appropriate value.
+
+In addition, we define the value `None`as a keyword to specify a parameter for
+which there should be no default value. The programmer must specifically
+provide a value for a parameter initialized with `None`, Otherwise, for an
+invocation that does not specify a value, the compiler may assume that all uses
```

Comment:
If the programmer must specifically provide a value for the parameter, how can there be an invocation that does not specify a value?

Reply:
If you use None, we assume that that parameter is not used at all. With a '_' (default value) we assume a default value described in the table (empty controls and parsers, default instances of externs, or zero valued ints, headers, structs, metadata.